src/main/java/org/openrewrite/staticanalysis/RemoveUnusedParams.java

```diff
@@ -0,0 +1,211 @@
+        return m.getBody() != null
+                && !m.hasModifier(J.Modifier.Type.Native)
+                && m.getLeadingAnnotations().isEmpty()
+                && !acc.contains(buildSignature(m));
+    }
+
+    private String buildSignature(final J.MethodDeclaration m) {
+        return m.getSimpleName() + "#"
+                + m.getMethodType().getParameterTypes().stream()
+                .map(Object::toString)
+                .collect(Collectors.joining(","));
+    }
+
+    private Set<String> collectUsedParameters(final J.MethodDeclaration m) {
+        Set<String> used = new HashSet<>();
+        Deque<Set<String>> shadowStack = new ArrayDeque<>();
+
+        new JavaIsoVisitor<Set<String>>() {
+            @Override
+            public J.Block visitBlock(final J.Block block, final Set<String> u) {
+                shadowStack.push(new HashSet<>());
+                try {
+                    return super.visitBlock(block, u);
+                } finally {
+                    shadowStack.pop();
+                }
+            }
+
+            @Override
+            public J.VariableDeclarations visitVariableDeclarations(final J.VariableDeclarations decl, final Set<String> u) {
+                decl.getVariables().forEach(v -> shadowStack.peek().add(v.getSimpleName()));
+                return super.visitVariableDeclarations(decl, u);
+            }
+
+            @Override
+            public J.Identifier visitIdentifier(final J.Identifier id, final Set<String> u) {
+                if (isVisibleParameter(id, m, shadowStack)) {
+                    u.add(id.getSimpleName());
+                }
+                return id;
+            }
+        }.visit(m.getBody(), used);
+
+        return used;
+    }
+
+    private boolean isVisibleParameter(final J.Identifier id, final J.MethodDeclaration m, final Deque<Set<String>> shadowStack) {
+        return !isShadowed(id.getSimpleName(), shadowStack) &&
+                isDeclaredAsParameter(id.getSimpleName(), m);
+    }
+
+    private boolean isShadowed(final String name, final Deque<Set<String>> shadowStack) {
+        return shadowStack.stream().anyMatch(scope -> scope.contains(name));
+    }
+
+    private boolean isDeclaredAsParameter(final String name, final J.MethodDeclaration m) {
+        return m.getParameters().stream()
+                .filter(p -> p instanceof J.VariableDeclarations)
+                .flatMap(p -> ((J.VariableDeclarations) p).getVariables().stream())
+                .anyMatch(v -> v.getSimpleName().equals(name));
+    }
+
+    private List<Statement> filterUnusedParameters(final J.MethodDeclaration m, final Set<String> usedParams) {
+        return m.getParameters().stream()
+                .flatMap(param -> {
+                    if (param instanceof J.VariableDeclarations) {
+                        J.VariableDeclarations decl = (J.VariableDeclarations) param;
+                        return filterDeclaration(decl, usedParams);
```

Comment:
```suggestion
                        return filterDeclaration(param, usedParams);
```
cast only needed for imperative style imho.

Reply:
```suggestion
                        return filterDeclaration((J.VariableDeclarations) param, usedParams);
```
param is of type Statement so i can pass it like this.

Reply:
yes then of course. thx.